src/python/pants/backend/native/subsystems/native_toolchain.py

```diff
@@ -60,14 +69,198 @@ def _llvm(self):
   def _xcode_cli_tools(self):
     return XCodeCLITools.scoped_instance(self)
 
-  @memoized_property
-  def _libc_dev(self):
-    return LibcDev.scoped_instance(self)
 
+class CompilerSystemDirSearchError(Exception):
+  """Thrown for errors in finding system includes and lib dirs for a compiler."""
+
+
+class DirCollectionRequest(datatype([('dir_paths', tuple)])): pass
+
+
+class ExistingDirCollection(datatype([('dirs', tuple)])): pass
+
+
+# FIXME: use snapshots for this!!!
+@rule(ExistingDirCollection, [Select(DirCollectionRequest)])
+def filter_existing_dirs(dir_collection_request):
+  real_dirs = OrderedSet()
+  for maybe_existing_dir in dir_collection_request.dir_paths:
+    # Could use a `seen_dir_paths` set if we want to avoid pinging the fs for duplicate entries.
+    if is_readable_dir(maybe_existing_dir):
+      real_dirs.add(os.path.realpath(maybe_existing_dir))
+    else:
+      logger.debug("found non-existent or non-accessible directory '{}' from request {}"
+                   .format(maybe_existing_dir, dir_collection_request))
+
+
+  return ExistingDirCollection(dirs=tuple(real_dirs))
+
+
+class CompilerSearchRequest(datatype([('compiler', SubclassesOf(CompilerMixin))])): pass
+
+
+class LibDirsFromCompiler(ExistingDirCollection): pass
+
+
+_search_dirs_libraries_regex = re.compile('^libraries: =(.*)$', flags=re.MULTILINE)
+
+
+@rule(LibDirsFromCompiler, [Select(CompilerSearchRequest)])
+def parse_known_lib_dirs(compiler_search_request):
+  # FIXME: convert this to using `copy()` when #6269 is merged.
+  cmplr = compiler_search_request.compiler
+  print_search_dirs_exe = type(cmplr)(path_entries=cmplr.path_entries,
```

Comment:
These should be tuples to conform to other instantiations in this changeset. I may be missing something, but its good to have consistency. Also, not sure how I feel about this construction using `type`. Might be cleaner to have this as a class method on the compiler class, however I don't have any explicit reason to change this beyond Python conventions/readability.

Reply:
+1 to classmethod over constructor here.

Reply:
I mentioned that this should be using a `.copy()` method with an explicit comment -- that has now been implemented and removes all of this.